{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a4d953c1-1fec-4f2a-a01d-9d4d5382bf60",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9aec4908-6056-4973-9445-c18d45c2d949",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(25426, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>token_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>nonzero_median</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&lt;pad&gt;</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&lt;mask&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>2</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>2.001186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>3</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>3.228213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>4</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>2.218874</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id  token_id gene_name  nonzero_median\n",
       "0            <pad>         0       NaN             NaN\n",
       "1           <mask>         1       NaN             NaN\n",
       "2  ENSG00000000003         2    TSPAN6        2.001186\n",
       "3  ENSG00000000005         3      TNMD        3.228213\n",
       "4  ENSG00000000419         4      DPM1        2.218874"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "token_path = \"/nfs/turbo/umms-indikar/shared/projects/geneformer/token_dictionary.pkl\"\n",
    "median_path = \"/nfs/turbo/umms-indikar/shared/projects/geneformer/geneformer/gene_median_dictionary.pkl\"\n",
    "id_path = \"/nfs/turbo/umms-indikar/shared/projects/geneformer/geneformer/gene_name_id_dict.pkl\"\n",
    "\n",
    "\n",
    "def make_gene_map(token_path, median_path, id_path):\n",
    "    \"\"\"\n",
    "    Combines gene dictionaries based on their keys.\n",
    "\n",
    "    Args:\n",
    "        token_path (str): Path to the token dictionary.\n",
    "        median_path (str): Path to the median dictionary.\n",
    "        id_path (str): Path to the gene name-to-ID dictionary.\n",
    "\n",
    "    Returns:\n",
    "        dict: A combined dictionary where keys are gene IDs and values are tuples containing:\n",
    "            - The tokenized representation of the gene.\n",
    "            - The median embedding of the gene.\n",
    "            - The gene name.\n",
    "    \"\"\"\n",
    "\n",
    "    with open(token_path, \"rb\") as f:\n",
    "        token_dict = pickle.load(f)\n",
    "\n",
    "    with open(median_path, \"rb\") as f:\n",
    "        median_dict = pickle.load(f)\n",
    "\n",
    "    with open(id_path, \"rb\") as f:\n",
    "        id_dict = pickle.load(f)\n",
    "        id_dict = {value: key for key, value in id_dict.items()}\n",
    "        \n",
    "    df = pd.DataFrame.from_dict(token_dict, orient='index')\n",
    "    df = df.reset_index()\n",
    "    df.columns = ['gene_id', 'token_id']\n",
    "    df['gene_name'] = df['gene_id'].map(id_dict)\n",
    "    df['nonzero_median'] = df['gene_id'].map(median_dict)\n",
    "    \n",
    "    return df\n",
    "\n",
    "df = make_gene_map(token_path, median_path, id_path)\n",
    "print(f\"{df.shape=}\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "18a1f938-30d9-4ba4-9eec-fce9cacce0af",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2663770/4104331383.py:3: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gtf = pd.read_csv(gtf_path)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gtf.shape=(41407, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>20</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>1471764</td>\n",
       "      <td>1497848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>629061</td>\n",
       "      <td>629433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>ENSG00000198744</td>\n",
       "      <td>5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>634375</td>\n",
       "      <td>634922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>ENSG00000279928</td>\n",
       "      <td>2</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>1</td>\n",
       "      <td>182695</td>\n",
       "      <td>184174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>ENSG00000142611</td>\n",
       "      <td>17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1</td>\n",
       "      <td>3069167</td>\n",
       "      <td>3438621</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id  gene_version            gene_biotype Chromosome  \\\n",
       "0    ENSG00000160072            20          protein_coding          1   \n",
       "111  ENSG00000225972             1  unprocessed_pseudogene          1   \n",
       "114  ENSG00000198744             5  unprocessed_pseudogene          1   \n",
       "117  ENSG00000279928             2  unprocessed_pseudogene          1   \n",
       "129  ENSG00000142611            17          protein_coding          1   \n",
       "\n",
       "       Start      End  \n",
       "0    1471764  1497848  \n",
       "111   629061   629433  \n",
       "114   634375   634922  \n",
       "117   182695   184174  \n",
       "129  3069167  3438621  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load GTF information \n",
    "gtf_path = \"/scratch/indikar_root/indikar1/cstansbu/HSC/references/geneTable.csv\"\n",
    "gtf = pd.read_csv(gtf_path)\n",
    "gtf = gtf[gtf['Feature'] == 'gene']\n",
    "gtf = gtf[gtf['gene_name'].notna()]\n",
    "\n",
    "columns = [ \n",
    "    'gene_id', \n",
    "    'gene_version',\n",
    "    # 'gene_name', \n",
    "    'gene_biotype',\n",
    "    'Chromosome', \n",
    "    'Start', \n",
    "    'End', \n",
    "]\n",
    "\n",
    "gtf = gtf[columns].drop_duplicates()\n",
    "print(f\"{gtf.shape=}\")\n",
    "gtf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "668dd5a2-4c32-4026-816f-9d0577f0b5e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1606\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['gene_name',\n",
       " 'ABL1',\n",
       " 'ACAA1',\n",
       " 'ADNP',\n",
       " 'ADNP2',\n",
       " 'AEBP2',\n",
       " 'AFF4',\n",
       " 'AHCTF1',\n",
       " 'AHDC1',\n",
       " 'AHR']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load scenic\n",
    "filepath = \"/nfs/turbo/umms-indikar/shared/projects/HSC/data/scenic_resources/500bp_up_100bp_down_B.csv\"\n",
    "sdf = pd.read_csv(filepath)\n",
    "sdf = sdf.rename(columns={\"Unnamed: 0\": \"gene_name\"})\n",
    "sdf.head()\n",
    "\n",
    "scenic_transcription_factors = sdf.columns.to_list()\n",
    "print(len(scenic_transcription_factors))\n",
    "scenic_transcription_factors[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d8f488a0-6025-46d4-8ade-eb5051d5af00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>token_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>nonzero_median</th>\n",
       "      <th>gene_version_x</th>\n",
       "      <th>gene_biotype_x</th>\n",
       "      <th>Chromosome_x</th>\n",
       "      <th>Start_x</th>\n",
       "      <th>End_x</th>\n",
       "      <th>scenic_tf</th>\n",
       "      <th>gene_version_y</th>\n",
       "      <th>gene_biotype_y</th>\n",
       "      <th>Chromosome_y</th>\n",
       "      <th>Start_y</th>\n",
       "      <th>End_y</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&lt;pad&gt;</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&lt;mask&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>2</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>2.001186</td>\n",
       "      <td>15.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>X</td>\n",
       "      <td>100627107.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>False</td>\n",
       "      <td>15.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>X</td>\n",
       "      <td>100627107.0</td>\n",
       "      <td>100639991.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>X</td>\n",
       "      <td>100627107.0</td>\n",
       "      <td>100639991.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>3</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>3.228213</td>\n",
       "      <td>6.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>X</td>\n",
       "      <td>100584935.0</td>\n",
       "      <td>100599885.0</td>\n",
       "      <td>False</td>\n",
       "      <td>6.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>X</td>\n",
       "      <td>100584935.0</td>\n",
       "      <td>100599885.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>X</td>\n",
       "      <td>100584935.0</td>\n",
       "      <td>100599885.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>4</td>\n",
       "      <td>DPM1</td>\n",
       "      <td>2.218874</td>\n",
       "      <td>14.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>20</td>\n",
       "      <td>50934866.0</td>\n",
       "      <td>50959140.0</td>\n",
       "      <td>False</td>\n",
       "      <td>14.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>20</td>\n",
       "      <td>50934866.0</td>\n",
       "      <td>50959140.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>20</td>\n",
       "      <td>50934866.0</td>\n",
       "      <td>50959140.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id  token_id gene_name  nonzero_median  gene_version_x  \\\n",
       "0            <pad>         0       NaN             NaN             NaN   \n",
       "1           <mask>         1       NaN             NaN             NaN   \n",
       "2  ENSG00000000003         2    TSPAN6        2.001186            15.0   \n",
       "3  ENSG00000000005         3      TNMD        3.228213             6.0   \n",
       "4  ENSG00000000419         4      DPM1        2.218874            14.0   \n",
       "\n",
       "   gene_biotype_x Chromosome_x      Start_x        End_x  scenic_tf  \\\n",
       "0             NaN          NaN          NaN          NaN      False   \n",
       "1             NaN          NaN          NaN          NaN      False   \n",
       "2  protein_coding            X  100627107.0  100639991.0      False   \n",
       "3  protein_coding            X  100584935.0  100599885.0      False   \n",
       "4  protein_coding           20   50934866.0   50959140.0      False   \n",
       "\n",
       "   gene_version_y  gene_biotype_y Chromosome_y      Start_y        End_y  \\\n",
       "0             NaN             NaN          NaN          NaN          NaN   \n",
       "1             NaN             NaN          NaN          NaN          NaN   \n",
       "2            15.0  protein_coding            X  100627107.0  100639991.0   \n",
       "3             6.0  protein_coding            X  100584935.0  100599885.0   \n",
       "4            14.0  protein_coding           20   50934866.0   50959140.0   \n",
       "\n",
       "   gene_version    gene_biotype Chromosome        Start          End  \n",
       "0           NaN             NaN        NaN          NaN          NaN  \n",
       "1           NaN             NaN        NaN          NaN          NaN  \n",
       "2          15.0  protein_coding          X  100627107.0  100639991.0  \n",
       "3           6.0  protein_coding          X  100584935.0  100599885.0  \n",
       "4          14.0  protein_coding         20   50934866.0   50959140.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge GTF information\n",
    "df = pd.merge(df, \n",
    "              gtf, \n",
    "              how='left',\n",
    "              left_on='gene_id',\n",
    "              right_on='gene_id',\n",
    "              )\n",
    "\n",
    "df['scenic_tf'] = df['gene_name'].isin(scenic_transcription_factors)\n",
    "\n",
    "# outpath = \"/nfs/turbo/umms-indikar/shared/projects/geneformer/token_mappings.csv\"\n",
    "outpath = \"../../data/token_mappings.csv\"\n",
    "df.to_csv(outpath, index=False)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "60bacacd-5b8c-4d8f-aad5-8c1254d50212",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (668683560.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[6], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    break\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4653e98-594b-4c5f-8d7f-8b44457a444e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "335fa71f-d148-4f26-b0da-fba6337dfebd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f4e3d75-58e1-47ad-8016-25d076382660",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f583ceee-3e34-44d5-8b54-a1672b1d08a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = [ \n",
    "    'gene_id', \n",
    "    'gene_version',\n",
    "    'gene_name', \n",
    "    'gene_biotype',\n",
    "    'Chromosome', \n",
    "    'Start', \n",
    "    'End', \n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5e11a71-cd14-4ca7-b007-9733dee0d576",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ec0528f-e898-45ae-9361-69a465025e98",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78ded0ec-15fd-4525-8a13-486b06ac3003",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87adc3ed-d37e-46d5-b499-2712c649ee92",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "geneformer",
   "language": "python",
   "name": "geneformer"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
