{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ecd14719-5572-4d73-b3e2-ab18d7a2b94b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import h5py\n",
    "import os\n",
    "import pickle\n",
    "import scipy.sparse as sp\n",
    "from geneformer import TranscriptomeTokenizer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d0604911-d406-4f2c-9aac-e7edf4d9342c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'control': 'control',\n",
       " 'data_id': 'data_id',\n",
       " 'dataset': 'dataset',\n",
       " 'hour': 'hour',\n",
       " 'n_counts': 'n_counts',\n",
       " 'sample_id': 'sample_id',\n",
       " 'timepoint': 'timepoint'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_path = \"/nfs/turbo/umms-indikar/shared/projects/geneformer/data/rajapakse_lab_data.h5ad\"\n",
    "input_dir = os.path.dirname(input_path)\n",
    "output_path = \"/scratch/indikar_root/indikar1/cstansbu/geneformer/\"\n",
    "prefix = \"test\"\n",
    "\n",
    "def get_attributes(h5ad_path):\n",
    "    \"\"\"\n",
    "    Extracts attribute names from the `.obs` field of an h5ad AnnData file,\n",
    "    returning them as a dictionary with keys and values being the attribute names.\n",
    "\n",
    "    Args:\n",
    "        h5ad_path (str): The path to the h5ad file.\n",
    "\n",
    "    Returns:\n",
    "        dict: A dictionary where keys and values are the unique attribute names \n",
    "              found in the `.obs` field of the h5ad file.\n",
    "    \"\"\"\n",
    "    with h5py.File(h5ad_path, mode=\"r\") as store:\n",
    "        attribute_names = list(store[\"obs\"].keys())\n",
    "\n",
    "    attribute_name_dict = {name: name for name in attribute_names}  # Create dictionary\n",
    "    return attribute_name_dict\n",
    "    \n",
    "\n",
    "custom_attr_name_dict = get_attributes(input_path)\n",
    "custom_attr_name_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "1d23082c-d593-4222-825b-13c2b6f9a15c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_gene_median_dict(gene_median_file):\n",
    "    \"\"\"\n",
    "    Loads a gene median dictionary from a pickle file.\n",
    "\n",
    "    Args:\n",
    "        gene_median_file (str): Path to the pickle file containing the gene median dictionary.\n",
    "\n",
    "    Returns:\n",
    "        dict: A dictionary mapping gene IDs to their median expression values.\n",
    "    \"\"\"\n",
    "\n",
    "    with open(gene_median_file, \"rb\") as f:\n",
    "        gene_median_dict = pickle.load(f)\n",
    "\n",
    "    return gene_median_dict\n",
    "\n",
    "\n",
    "def load_gene_tokenization(token_dictionary_file):\n",
    "    \"\"\"\n",
    "    Loads gene tokenization data from a pickle file.\n",
    "\n",
    "    Args:\n",
    "        token_dictionary_file (str): Path to the pickle file containing the gene-token dictionary.\n",
    "\n",
    "    Returns:\n",
    "        dict: Gene-token dictionary (Ensembl ID: token).\n",
    "        list: List of all gene keys (Ensembl IDs).\n",
    "        dict: Dictionary mapping gene keys to True (used for selecting genes later).\n",
    "    \"\"\"\n",
    "\n",
    "    with open(token_dictionary_file, \"rb\") as f:\n",
    "        gene_token_dict = pickle.load(f)\n",
    "\n",
    "    gene_keys = list(gene_token_dict.keys())\n",
    "\n",
    "    # Optimization: Pre-allocate the list for slight performance improvement\n",
    "    genelist_dict = dict.fromkeys(gene_keys, True)\n",
    "\n",
    "    return gene_token_dict, gene_keys, genelist_dict\n",
    "\n",
    "\n",
    "def rank_genes(gene_vector, gene_tokens):\n",
    "    \"\"\"Ranks genes based on expression values in descending order.\n",
    "\n",
    "    Args:\n",
    "        gene_vector (numpy.ndarray): Array of gene expression values.\n",
    "        gene_tokens (numpy.ndarray): Array of corresponding gene tokens.\n",
    "\n",
    "    Returns:\n",
    "        numpy.ndarray: Array of gene tokens sorted by descending expression value.\n",
    "    \"\"\"\n",
    "    return gene_tokens[np.argsort(-gene_vector)]\n",
    "\n",
    "\n",
    "def normalize_counts(adata_chunk,  counts_column='n_counts', target_sum=10000):\n",
    "    \"\"\"Normalizes gene expression counts within a chunk of AnnData.\n",
    "\n",
    "    Args:\n",
    "        adata_chunk (AnnData): A chunk of the AnnData object containing gene expression data.\n",
    "        counts_column (str): Name of the column in `adata_chunk.obs` containing the total counts per cell.\n",
    "        target_sum (float): The desired total count per cell after normalization.\n",
    "        norm_factor_vector (numpy.ndarray): An array of normalization factors for each gene.\n",
    "\n",
    "    Returns:\n",
    "        scipy.sparse.csr_matrix: A sparse matrix containing the normalized gene expression counts.\n",
    "\n",
    "    This function performs the following steps:\n",
    "        1. Extracts the total counts per cell from the specified column (`counts_column`).\n",
    "        2. Normalizes the gene expression matrix (`adata_chunk.X`) by dividing by the total counts \n",
    "           and multiplying by the `target_sum`.\n",
    "        3. Further adjusts the normalized values by dividing by the gene-specific normalization \n",
    "           factors (`norm_factor_vector`).\n",
    "        4. Returns the normalized expression matrix as a sparse CSR matrix for efficient storage \n",
    "           and computation.\n",
    "    \"\"\"\n",
    "    \n",
    "    n_counts = adata_chunk.obs[counts_column].values[:, None]  # Cell counts as column vector\n",
    "    X_norm = adata_chunk.X / n_counts * target_sum / norm_factor_vector\n",
    "    return sp.csr_matrix(X_norm)  # Efficient sparse representation\n",
    "\n",
    "\n",
    "def tokenize_anndata(adata, genelist_dict, gene_median_dict, \n",
    "                     chunk_size=100000, target_sum=10000):\n",
    "    \"\"\"\n",
    "    Tokenizes and ranks genes within an AnnData object, optimizing for memory efficiency.\n",
    "\n",
    "    This function processes gene expression data in chunks, applies normalization, and ranks genes\n",
    "    for each cell based on their expression levels. The resulting tokenized and ranked gene\n",
    "    representations, along with cell metadata, are returned.\n",
    "\n",
    "    Args:\n",
    "        adata (AnnData): The AnnData object containing gene expression data.\n",
    "        genelist_dict (dict): Dictionary mapping gene IDs to boolean values indicating relevance.\n",
    "        gene_median_dict (dict): Dictionary mapping gene IDs to their median expression values.\n",
    "        chunk_size (int, optional): Number of cells to process in each chunk (default: 1000).\n",
    "        target_sum (int, optional): Target sum for count normalization (default: 10000).\n",
    "\n",
    "    Returns:\n",
    "        tuple: \n",
    "            - list: List of tokenized and ranked gene lists for each cell.\n",
    "            - dict: Dictionary containing cell metadata (keys are metadata column names).\n",
    "    \"\"\"\n",
    "    # Filter relevant miRNAs\n",
    "    coding_miRNA_mask = np.array([genelist_dict.get(i, False) for i in adata.var['ensembl_id']])\n",
    "    coding_miRNA_loc = np.where(coding_miRNA_mask)[0]\n",
    "\n",
    "    # Extract miRNA information\n",
    "    coding_miRNA_ids = adata.var['ensembl_id'][coding_miRNA_loc]\n",
    "    norm_factor_vector = np.array([gene_median_dict[i] for i in coding_miRNA_ids])\n",
    "    coding_miRNA_tokens = np.array([gene_token_dict[i] for i in coding_miRNA_ids])\n",
    "\n",
    "    tokenized_cells = []\n",
    "    file_cell_metadata = {k: [] for k in adata.obs.columns}  # Initialize metadata dict\n",
    "\n",
    "    # Process in chunks for memory efficiency\n",
    "    for chunk_start in range(0, adata.shape[0], chunk_size):\n",
    "        chunk_end = chunk_start + chunk_size\n",
    "        adata_chunk = adata[chunk_start:chunk_end, coding_miRNA_loc]\n",
    "        \n",
    "        # Normalize counts (could be replaced with the untested function above)\n",
    "        n_counts = adata_chunk.obs['n_counts'].values[:, None]\n",
    "        X_norm = adata_chunk.X / n_counts * target_sum / norm_factor_vector\n",
    "        X_norm = sp.csr_matrix(X_norm)  \n",
    "\n",
    "        # Tokenize and rank genes for each cell in chunk\n",
    "        for i in range(X_norm.shape[0]):\n",
    "            ranks = rank_genes(X_norm[i].data, coding_miRNA_tokens[X_norm[i].indices])\n",
    "            ranks = list(ranks[~np.isnan(ranks)].astype(int))\n",
    "\n",
    "            tokenized_cells.append(ranks)\n",
    "\n",
    "        # Update metadata\n",
    "        for k in adata.obs.columns:\n",
    "            file_cell_metadata[k].extend(adata_chunk.obs[k].tolist())\n",
    "\n",
    "    return tokenized_cells, file_cell_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "26a15cbf-dab0-440f-a003-d4b6801011ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "DEFAULT_TOKEN_PATH = \"/nfs/turbo/umms-indikar/shared/projects/geneformer/token_dictionary.pkl\"\n",
    "DEFAULT_MEDIAN_PATH = \"/nfs/turbo/umms-indikar/shared/projects/geneformer/geneformer/gene_median_dictionary.pkl\"\n",
    "\n",
    "gene_token_dict, gene_keys, genelist_dict = load_gene_tokenization(DEFAULT_TOKEN_PATH)\n",
    "gene_median_dict = load_gene_median_dict(DEFAULT_MEDIAN_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c0eed918-a3b6-486d-827b-86eb9681c491",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3347022/858760088.py:108: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`\n",
      "  coding_miRNA_ids = adata.var['ensembl_id'][coding_miRNA_loc]\n"
     ]
    }
   ],
   "source": [
    "tokenized_cells, cell_metadata = tokenize_anndata(adata, \n",
    "                                                  genelist_dict, \n",
    "                                                  gene_median_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "87576d31-6ed7-4bf9-bcd3-08b9b1c73aa7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 66 × 19393 backed at '/nfs/turbo/umms-indikar/shared/projects/geneformer/data/rajapakse_lab_data.h5ad'\n",
       "    obs: 'dataset', 'sample_id', 'timepoint', 'hour', 'n_counts', 'control'\n",
       "    var: 'gene_id', 'token_id', 'Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand', 'Frame', 'gene_version', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'ccds_id', 'exon_number', 'exon_id', 'exon_version', 'protein_id', 'protein_version', 'transcript_support_level', 'ensembl_id'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4939108c-6b4a-4ee3-b233-017a1d44e8cd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36b3b5a6-dcb4-40ce-b063-73b761ac4c33",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bab922fd-79b5-4aca-9ac1-71af90dacee9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b3609b6-cea8-4a32-91e6-3f4efbf1d62a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a259492d-e9c7-4a42-8fb1-f71d0a54aed5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tokenize_anndata(adata_file_path, target_sum=10_000, custom_attr_name_dict=None):\n",
    "    adata = ad.read(adata_file_path, backed=\"r\")\n",
    "\n",
    "    if custom_attr_name_dict is not None:\n",
    "        file_cell_metadata = {\n",
    "            attr_key: [] for attr_key in custom_attr_name_dict.keys()\n",
    "        }\n",
    "\n",
    "    coding_miRNA_loc = np.where(\n",
    "        [self.genelist_dict.get(i, False) for i in adata.var[\"ensembl_id\"]]\n",
    "    )[0]\n",
    "    norm_factor_vector = np.array(\n",
    "        [\n",
    "            gene_median_dict[i]\n",
    "            for i in adata.var[\"ensembl_id\"][coding_miRNA_loc]\n",
    "        ]\n",
    "    )\n",
    "    coding_miRNA_ids = adata.var[\"ensembl_id\"][coding_miRNA_loc]\n",
    "    coding_miRNA_tokens = np.array(\n",
    "        [self.gene_token_dict[i] for i in coding_miRNA_ids]\n",
    "    )\n",
    "\n",
    "    try:\n",
    "        _ = adata.obs[\"filter_pass\"]\n",
    "    except KeyError:\n",
    "        var_exists = False\n",
    "    else:\n",
    "        var_exists = True\n",
    "\n",
    "    if var_exists:\n",
    "        filter_pass_loc = np.where([i == 1 for i in adata.obs[\"filter_pass\"]])[0]\n",
    "    elif not var_exists:\n",
    "        print(\n",
    "            f\"{adata_file_path} has no column attribute 'filter_pass'; tokenizing all cells.\"\n",
    "        )\n",
    "        filter_pass_loc = np.array([i for i in range(adata.shape[0])])\n",
    "\n",
    "    tokenized_cells = []\n",
    "\n",
    "    for i in range(0, len(filter_pass_loc), self.chunk_size):\n",
    "        idx = filter_pass_loc[i : i + self.chunk_size]\n",
    "\n",
    "        n_counts = adata[idx].obs[\"n_counts\"].values[:, None]\n",
    "        X_view = adata[idx, coding_miRNA_loc].X\n",
    "        X_norm = X_view / n_counts * target_sum / norm_factor_vector\n",
    "        X_norm = sp.csr_matrix(X_norm)\n",
    "\n",
    "        tokenized_cells += [\n",
    "            rank_genes(X_norm[i].data, coding_miRNA_tokens[X_norm[i].indices])\n",
    "            for i in range(X_norm.shape[0])\n",
    "        ]\n",
    "\n",
    "        # add custom attributes for subview to dict\n",
    "        if self.custom_attr_name_dict is not None:\n",
    "            for k in file_cell_metadata.keys():\n",
    "                file_cell_metadata[k] += adata[idx].obs[k].tolist()\n",
    "        else:\n",
    "            file_cell_metadata = None\n",
    "\n",
    "    return tokenized_cells, file_cell_metadata\n",
    "\n",
    "\n",
    "tokenize_anndata(input_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b66c5eba-9af0-478c-81bc-483b203a6830",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a712c85-e885-4b7f-a5b4-e1e576155426",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34549d59-bab4-4fc4-bb08-5230e0343617",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab2134e5-4128-417e-8680-f678821314f0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5e2a7c81-2145-4b6b-82b3-ba61577eda38",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tk.tokenize_data(input_dir, \n",
    "#                  output_path, \n",
    "#                  prefix, \n",
    "#                  file_format=\"h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "44c74a2f-5190-4d9f-882e-c55ec9db94ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/cstansbu/miniconda3/envs/geneformer/lib/python3.10/site-packages/anndata/__init__.py:55: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 66 × 19393 backed at '/nfs/turbo/umms-indikar/shared/projects/geneformer/data/rajapakse_lab_data.h5ad'\n",
       "    obs: 'dataset', 'sample_id', 'timepoint', 'hour', 'n_counts', 'control'\n",
       "    var: 'gene_id', 'token_id', 'Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand', 'Frame', 'gene_version', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'ccds_id', 'exon_number', 'exon_id', 'exon_version', 'protein_id', 'protein_version', 'transcript_support_level', 'ensembl_id'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = ad.read(input_path, backed=\"r\")\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ef3ead52-25bd-4bc0-8718-91c24230efab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>token_id</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>ENSG00000121410</td>\n",
       "      <td>5150.0</td>\n",
       "      <td>19</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>58345177.0</td>\n",
       "      <td>58353492.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000121410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>ENSG00000148584</td>\n",
       "      <td>9064.0</td>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>50799408.0</td>\n",
       "      <td>50885675.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000148584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>ENSG00000175899</td>\n",
       "      <td>13826.0</td>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9067663.0</td>\n",
       "      <td>9116229.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000175899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>ENSG00000166535</td>\n",
       "      <td>11812.0</td>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8822620.0</td>\n",
       "      <td>8887001.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000166535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A3GALT2</th>\n",
       "      <td>ENSG00000184389</td>\n",
       "      <td>15327.0</td>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33306765.0</td>\n",
       "      <td>33321098.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000184389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_id  token_id Chromosome          Source Feature  \\\n",
       "gene_name                                                                 \n",
       "A1BG       ENSG00000121410    5150.0         19  ensembl_havana    gene   \n",
       "A1CF       ENSG00000148584    9064.0         10  ensembl_havana    gene   \n",
       "A2M        ENSG00000175899   13826.0         12  ensembl_havana    gene   \n",
       "A2ML1      ENSG00000166535   11812.0         12  ensembl_havana    gene   \n",
       "A3GALT2    ENSG00000184389   15327.0          1  ensembl_havana    gene   \n",
       "\n",
       "                Start         End Score Strand Frame  ...  transcript_biotype  \\\n",
       "gene_name                                             ...                       \n",
       "A1BG       58345177.0  58353492.0     .      -     .  ...                 NaN   \n",
       "A1CF       50799408.0  50885675.0     .      -     .  ...                 NaN   \n",
       "A2M         9067663.0   9116229.0     .      -     .  ...                 NaN   \n",
       "A2ML1       8822620.0   8887001.0     .      +     .  ...                 NaN   \n",
       "A3GALT2    33306765.0  33321098.0     .      -     .  ...                 NaN   \n",
       "\n",
       "          tag ccds_id  exon_number  exon_id  exon_version  protein_id  \\\n",
       "gene_name                                                               \n",
       "A1BG      NaN     NaN          NaN      NaN           NaN         NaN   \n",
       "A1CF      NaN     NaN          NaN      NaN           NaN         NaN   \n",
       "A2M       NaN     NaN          NaN      NaN           NaN         NaN   \n",
       "A2ML1     NaN     NaN          NaN      NaN           NaN         NaN   \n",
       "A3GALT2   NaN     NaN          NaN      NaN           NaN         NaN   \n",
       "\n",
       "           protein_version  transcript_support_level       ensembl_id  \n",
       "gene_name                                                              \n",
       "A1BG                   NaN                       NaN  ENSG00000121410  \n",
       "A1CF                   NaN                       NaN  ENSG00000148584  \n",
       "A2M                    NaN                       NaN  ENSG00000175899  \n",
       "A2ML1                  NaN                       NaN  ENSG00000166535  \n",
       "A3GALT2                NaN                       NaN  ENSG00000184389  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85254dbd-de73-4a8d-9216-599a5e3ea06b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "65721086-af50-4a98-bd1a-296b4ab9aa39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 2 × 2 backed at '/nfs/turbo/umms-indikar/shared/projects/geneformer/data/rajapakse_lab_data.h5ad'\n",
       "    obs: 'dataset', 'sample_id', 'timepoint', 'hour', 'n_counts', 'control'\n",
       "    var: 'gene_id', 'token_id', 'Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand', 'Frame', 'gene_version', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'ccds_id', 'exon_number', 'exon_id', 'exon_version', 'protein_id', 'protein_version', 'transcript_support_level', 'ensembl_id'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx = ['63246_T0R1', '63249_T2R1']\n",
    "var_idx = ['A1BG', 'A2M']\n",
    "\n",
    "X_view = adata[idx, var_idx]\n",
    "X_view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d87bbf69-9b21-4f7f-86ab-92ec81d6a60a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "geneformer",
   "language": "python",
   "name": "geneformer"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
